feature/ATC-176

erikquinn opened the pull request:
Restores functionality of the video maps, and does a moderate amount of reorganizing and simplification of the handling of coordinates when instantiating a new `PositionModel`.

erikquinn committed:
Fix minor spelling error

erikquinn committed:
Fix PositionModel.calculatePosition static method

erikquinn committed:
Consolidate duplicate functions for calculating bearing to a point

erikquinn committed:
Simplify coordinate parsing process and organization

erikquinn commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -267,49 +247,26 @@ export default class PositionModel {
-        rawX = coordinates[LATITUDE_INDEX];
-        rawY = coordinates[LONGITUDE_INDEX];
+        throw new TypeError(`Invalid parameter. PositionModel.getPosition() requires
+                            coordinates, referencePostion and magneticNorth as parameters`);
```
Asked about this in slack as well... is this an acceptable resolution to the lint warning due to length of the line?

n8rzz replied:
in some cases, yes.  like boolean logic and such.  Strings are a little different, especially template strings (which this is).

there are two possible solutions:
```javascript
throw new TypeError(`Invalid parameter. PositionModel.getPosition() requires ` +
    `coordinates, referencePostion and magneticNorth as parameters`);
```
```javascript
// eslint-disable-line max-len
throw new TypeError(`Invalid parameter. PositionModel.getPosition() requires coordinates, referencePostion and magneticNorth as parameters`);
```
with error messaging, I tend to favor the later (with or without the eslint comment).

n8rzz commented on src/assets/scripts/airport/Arrival/ArrivalBase.js:
```diff
@@ -405,7 +405,7 @@ export default class ArrivalBase {
                 waypoints: this.fixes,
                 route: _get(this, 'activeRouteModel.routeString', ''),
-                position: aircraftPosition.position,
+                position: pos.position,
```
👍 nice work moving this around.  I like this a lot better, it's much much cleaner.

n8rzz commented on src/assets/scripts/airport/Arrival/ArrivalBase.js:
```diff
@@ -15,7 +15,7 @@ import {
     isWithinAirspace,
     calculateDistanceToBoundary,
-    calculateHeadingFromTwoPositions
+    bearingToPoint
```
:+1 good name choice

n8rzz commented on src/assets/scripts/airport/Arrival/ArrivalBase.js:
```diff
@@ -360,13 +360,14 @@ export default class ArrivalBase {
-                    const heading = bearing(previousFix.gps, nextFix.gps);
-                    const calculatedRadialDistance = fixRadialDist(previousFix.gps, heading, spawnOffset);
+                    const heading = bearingToPoint(previousFix.gps, nextFix.gps);
+                    const spawnPoint = fixRadialDist(previousFix.gps, heading, spawnOffset).reverse();
```
what is the intent with `.reverse()` here?  why does this need to be reversed?  

please add a comment here

erikquinn replied:
Received `[longitude, latitude]`, because `PositionModel.gps` returns in [x,y] order, and expected `[latitude, longitude]`. Addressed this in recent revisions to the PR.

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -66,7 +66,7 @@ export default class PositionModel {
      * @param mode {string}                     Set to 'GPS' to indicate you are inputting lat/long that should
      *                                          be converted to positions
      */
-    constructor(coordinates = [], reference, magnetic_north = 0, mode) {
```
did you check to see if this is still in use?  I seem to remember at least one case where a new PositionModel was sending this mode parameter.

erikquinn replied:
It was still in use, but it didn't need to be. I've reworked the thing that was using it, so the mode is no longer needed. :smile:

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -183,43 +172,7 @@ export default class PositionModel {
-            this.reference_position.longitude
-        );
-
-        this.adjustXYForMagneticNorth();
```
👍  Yay refactoring!

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -243,14 +196,41 @@ export default class PositionModel {
+        );
+
+        this._adjustXYForMagneticNorth();
+    }
```
This might blow your mind.  Now that you've made the `.calculatePosition()` method much easier to work with, instead of repeating that (minimal) logic here, you can actually call the static method:

```javascript
const [ x, y ] = PositionModel.calculatePosition([this.latitude, this.longitude], this.reference_position, this.magnetic_north);
 
this.x = x;
this.y = y;
```

that also means that we can remove the `_adjustXYForMagneticNorth()` method entirely.

erikquinn replied:
Fixed. :+1:

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -267,49 +247,26 @@ export default class PositionModel {
-    latitude = rawY;
-
-    rawX = calculateDistanceToPointForX(
+    const canvasPositionX = calculateDistanceToPointForX(
```
👍  good name

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -267,49 +247,26 @@ export default class PositionModel {
-    }
-
-    rawY = calculateDistanceToPointForY(
+    const canvasPositionY = calculateDistanceToPointForY(
```
👍  good name

n8rzz commented on src/assets/scripts/math/flightMath.js:
```diff
@@ -54,14 +54,14 @@ export const calcTurnInitiationDistance = (speed, bankAngle, courseChange) => {
-export const bearing = (startPosition, endPosition) => {
+export const bearingToPoint = (startPosition, endPosition) => {
     return vradial(vsub(endPosition, startPosition));
 };
```
I like how you moved the parameters around here. I believe we discussed this method a while back, it makes more sense to have start then end.

you could still use an implicit return with this function and keep it to one line:
```javascript
export const bearingToPoint = (startPosition, endPosition) => vradial(vsub(endPosition, startPosition));
```
That and what you have are exactly the same thing, though, just slightly different syntax.

erikquinn replied:
Fixed. :+1:

n8rzz commented on src/assets/scripts/utilities/unitConverters.js:
```diff
@@ -220,30 +220,35 @@ export const heading_to_string = (heading) => {
  * @return transformedCoordinate {number}
  */
-export const parseCoordinate = (coordintae) => {
-    const match = REGEX.LAT_LONG.exec(coordintae);
```
spelling is hard :-(

n8rzz commented on src/assets/scripts/utilities/unitConverters.js:
```diff
@@ -220,30 +220,35 @@ export const heading_to_string = (heading) => {
+export const parseCoordinate = (coordinate) => {
+    const match = REGEX.LAT_LONG.exec(coordinate);
 
+    // If coordinate already in WGS84 ESPG:4326 form ('39.427618, -75.296011'), just return it as-is
```
👍 for the comment here

n8rzz commented on src/assets/scripts/utilities/unitConverters.js:
```diff
@@ -220,30 +220,35 @@ export const heading_to_string = (heading) => {
+    // Gather minutes/seconds as decimal of a degree, if available
     if (match[5] != null) {
-        transformedCoordinate += parseFloat(match[5]) / 60;
+        minutes = (parseFloat(match[5]) * TIME.ONE_MINUTE_IN_HOURS);
```
this certainly reads better.  One day I'd like to enumerate all the indicies this regex.match throws out.

n8rzz commented on src/assets/scripts/utilities/unitConverters.js:
```diff
@@ -220,30 +220,35 @@ export const heading_to_string = (heading) => {
+    // Apply negative to value if coordinate is "South" or "West"
     if (REGEX.SW.test(match[1])) {
-        transformedCoordinate *= -1;
+        decimalDegrees *= -1;
```
👍 good name

n8rzz commented on test/math/flightMath.spec.js:
```diff
@@ -26,11 +26,11 @@ ava('.calcTurnRadius() returns a turn radius based on speed and bank angle', t =
 });
 
-ava('.calculateHeadingFromTwoPositions()', t => {
+ava('.bearingToPoint()', t => {
```
I won't knock you on the lack of description here because, well, I was naughty and didn't have one there first.  In the future, though, we should have descriptions for our tests.  I'm very guilty here, but something to keep in mind.

n8rzz reviewed (changes requested):
This looks wonderful!  I really like how the PositionModel is _so_ much simpler and easier to read than it was a month ago.  Good work.

Have you have a chance to test this in the browser?  I ask because for me, locally, I don't see any terrain maps with these new changes.

I have an issue out there for developing a test plan ( #131 ). That is going to be just a list of things that are expected to work.  So when we make big changes we can just go down the list and be confident everything is in working order.

erikquinn committed:
Change `PositionModel.gps` getter to [lat,lon] order
- and replace all references to the former `PositionModel.gps` that require the [x,y] order to a new getter called `PositionModel.gpsXY`

erikquinn committed:
Simplify PositionModel._calculateScreenPosition by using static calculatePosition method

erikquinn commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -238,15 +204,13 @@ export default class PositionModel {
-    adjustXYForMagneticNorth() {
-        const { x, y } = adjustForMagneticNorth(this.x, this.y, this.magnetic_north);
+    _calculateScreenPosition() {
+        const [x, y] = PositionModel.calculatePosition(this.gps, this.reference_position, this.magnetic_north);
```
This call to `PositionModel.calculatePosition()` fails when its the airport itself trying to create its `PositionModel`. Since there is no reference position (everything will be in reference to the airport), the method throws the error in the first line:
```javascript
    if (!coordinates || !referencePostion || !magneticNorth) {
        throw new TypeError('Invalid parameter. PositionModel.getPosition() requires coordinates, referencePostion ' +
            'and magneticNorth as parameters');
    }
```

Any ideas on how this might best be resolved?

n8rzz replied:
ahh yes, there should probably be an early out before this throws to handle that case, or maybe before it gets to this function.  

perhaps this is handled in the `_calculateScreenPosition()` method where we do:
```javascript
_calculateScreenPosition() {
    if (!this.referencePosition) {
        return;
    }
    // ...
}
```

n8rzz replied:
theoretically, anything calling that static method _should_ already have a referencePosition.

erikquinn reviewed (commented):
@n8rzz This stops the error.

n8rzz commented on src/assets/scripts/airport/AirportModel.js:
```diff
@@ -180,7 +180,7 @@ export default class AirportModel {
         }
 
-        this.position = new PositionModel(currentPosition, null, magneticNorth);
+        this.position = new PositionModel(currentPosition, currentPosition, magneticNorth);
```
To be honest, I don't think this the right solution.  It may have solved that current problem, but this creates a few more.  I still think an early exit is going to be the best solution when a `referencePosition` is not present.

Within the `PositionModel` we have:
```javascript
/**
 * @property reference_position
 * @type {PositionModel|null}
  */
this.reference_position = reference;
```
Where `reference_position` is a `PositionModel` or `null`.  Passing `currentPosition` as the second argument now introduces a third possible type for `reference_position`.  That is not a good idea because later on, you might be expecting to deal with a `PositionModel` and instead be getting a simple array.  Then things break.  

It is _sooo_ much easier for properties to be a single type.  I should mention that `PositionModel|null` is still considered the same type because a class is either an instanceof a class or null.

n8rzz reviewed (changes requested):
Make sure to run the tests.  We have a growing test suit, 170+, and they are there to help make sure things aren't breaking.  When I pulled this branch and ran locally, I only had 58 pass and 12 uncaught exceptions.  Part of that is because the `PositionModel` is used _everywhere_ so one typo could cause a lot of breakage.  But it also might have to do with the heavy refactor that class has experienced.

Also, when I run this branch locally I am still not seeing any terrain maps.

erikquinn committed:
Change Position construction to only use refPos if available

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -238,15 +204,28 @@ export default class PositionModel {
-    adjustXYForMagneticNorth() {
-        const { x, y } = adjustForMagneticNorth(this.x, this.y, this.magnetic_north);
+    _calculateScreenPosition() {
+        if (!this._hasReferencePosition()) {
```
this same logic could be done with `!this.reference_position`, but I like this abstraction be cause it reads better. 👍

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -238,15 +204,28 @@ export default class PositionModel {
-     *
-     * Adjusts x & y coordinates from true north to magnetic north
-     *
+     * Checks whether or not this `PositionModel` has a reference PositionModel
```
👍 for the comments! (be sure to remember the bat ticks on the second `PositionModel` reference though)

erikquinn replied:
Fixed. 👍

n8rzz reviewed (commented):
This looks way better!

For the failing test, instead of `t.notThrows()` it should probably be `t.throws()`.  The terrain maps are still failing to render but that appears to be a mathing error.  example:
`atc-dev` klas `AirportModel.terrain.1000[0][0][0]` (the first array value in the terrain object) = `
[50.78703663558661, -31.587506650161064]`

in this branch it is `[14655.424074276078, -14042.615087087366]`

erikquinn committed:
Amend PositionModel tests to pass

erikquinn committed:
Restore terrain

n8rzz commented on src/assets/scripts/airport/Arrival/ArrivalBase.js:
```diff
@@ -360,13 +360,14 @@ export default class ArrivalBase {
-                    const calculatedRadialDistance = fixRadialDist(previousFix.gps, heading, spawnOffset);
+                    const heading = bearingToPoint(previousFix.gpsXY, nextFix.gpsXY);
+                    const spawnPoint = fixRadialDist(previousFix.gps, heading, spawnOffset);
+                    const spawnPosition = new PositionModel(spawnPoint, this.airport.position, this.airport.magnetic_north);
```
something in this area broke initial spawn positions such that the initially spawned aircraft are not visible.  They are created, but never show up in the view.

erikquinn replied:
Oh, of course!!! :stuck_out_tongue_winking_eye: 

Had to do with the `fixRadialDist` function expecting [x,y] order, despite the function description indicating otherwise. Fixed it to deal properly with [lat,lon] order instead. Arrivals pre-spawn properly again.

n8rzz reviewed (changes requested):
Looking really good!  

See if you can narrow down that spawn position issue, though.  I narrowed it down to your commit from 12/2 (504794c755992c6995a30a0466da30428762b2eb).  I'm sure you'll look at it and be like, 'Oh of course!'.

erikquinn committed:
Fix to `fixRadialDist` function
- Restores pre-spawned aircraft

erikquinn committed:
Add test for `fixRadialDist` function

n8rzz commented on src/assets/scripts/base/PositionModel.js:
```diff
@@ -267,49 +246,26 @@ export default class PositionModel {
     if (!coordinates || !referencePostion || !magneticNorth) {
-        throw new TypeError('Invalid parameter. PositionModel.getPosition() requires coordinates, referencePostion and magneticNorth as parameters');
+        throw new TypeError('Invalid parameter. PositionModel.getPosition() requires coordinates, referencePostion ' +
+            'and magneticNorth as parameters');
```
We talked about this briefly over slack.  When it comes to error messages, I'm ok with exceeding the min line length, as long as they aren't too long, and adding an `// eslint-ignore-line max-len` comment.  

You are correct to break lines in this case, however. 👍

n8rzz commented on test/math/flightMath.spec.js:
```diff
@@ -26,11 +27,22 @@ ava('.calcTurnRadius() returns a turn radius based on speed and bank angle', t =
+    const result = fixRadialDist(positionStart, radial, distance_nm);
+
+    t.true(result[0] === expectedResult[0]);
+    t.true(result[1] === expectedResult[1]);
```
👍 👍 👍 
Look at you, adding tests.  Hell yeah!!

_fun fact_: When I'm comparing arrays or objects within tests, sometimes I use `_isEqual()` to help me out.  You did good with this test, I like the verbosity.  I just want to make you aware of it because there may come a time when you need to compare larger arrays/objects.   As an example, using your test above: 
```javascript
t.true(_isEqual(result, expectedResult));
```

You don't need to change anything, I like this test the way it is.  Just something to keep in mind if you run into that.  Bringing in lodash functions to help in tests is totally cool.

n8rzz reviewed (approved):
This looks good!  Great work!!  

Thanks for the hard work on this one.  You're making good progress with your javascripts and I'd like to see that continue.  Keep it up!

n8rzz committed:
feature/ATC-176 - Updates changelog
